platform-sdk/docs/proposals/virtualmegamap-binary-merkledb/proposal.md

```diff
@@ -0,0 +1,162 @@
+Application code has to be changed to provide the serializers to `VirtualMap` rather than to MerkleDb.
+
+`KeySerializer` and `ValueSerializer` interfaces are moved from one package to another, but all
+interface methods are the same. Application code needs to be updated to use the new package (and
+another Java module).
+
+## Compatibility and Migration
+
+No data migration is expected at this point. Data on disk is stored in the same format (100% byte to
+byte compatibility) as previously.
+
+When a new virtual map is created, it gets key and value serializers from the app. However, there
+is a number of already created maps in the old states. They are not created from scratch, but loaded
+from state snapshots. Where do they get the serializers from? MerkleDb. Despite MerkleDb no longer
+uses serializers, it still loads them from disk (MerkleDb configuration), if available. When a
+virtual map is loaded from a state snapshot, and snapshot is old (before MerkleDb is switched to
+work with bytes), the map will query the data source for key and value serializers to use.
+
+## Testing
+
+This feature doesn't add any new functionality. There is no need to create new unit tests. The
+existing unit tests must be updated according to API and class changes above:
+
+* `VirtualDataSource` interface methods are changed to use bytes
+* `VirtualDataSourceBuilder` interface methods are changed
+* Many MerkleDb classes are changed to use bytes
+* When a virtual map is created, it now needs key and value serializers. Previously it was only
+  needed, if the map used MerkleDb data source, but not other data sources, e.g.
+  `InMemoryDataSource`
+
+MerkleDb unit tests are affected the most. Majority of virtual map unit tests require little to
+no updates.
+
+### Migration testing
+
+No changes are expected to data files on disk. This has to be tested thoroughly. Existing state
+snapshots, e.g. from mainnet, from old software versions can be used to start a node with
+MerkleDb changes. The node must be able to start fine, and all data from the snapshots must be
+accessible.
+
+## Related changes
+
+State validation tool uses some MerkleDb classes to validate data files on disk. Some of these
+classes will be changed to use bytes rather than objects. Validation tool sources should be
+updated accordingly.
+
+## Performance
+
+Performance impact has to be measured carefully. There are no flow changes in how data is stored
+to disk or read back. However, there are changes to when and how virtual keys and values are
+converted to bytes and vice versa.
+
+Previously, during flush, an entity (virtual key and value) was stored in the node cache as a
+Java object, passed to MerkleDb, and written to disk using key and value serializers. The
+serializers wrote key and value bytes directly to the file. With the new approach, entities
+are serialized to bytes before flush. It results in extra Java byte array allocations. Then
+these byte arrays are written to the files.
```

Comment:
The extra heap usage seems a bit concerning. I'm wondering if there's a solution that would still stream the bytes directly from the serializer to the disk, perhaps with a tiny proxy in between, but w/o having to allocate byte arrays for every entity. The latter seems a bit wasteful.

I'm curious to see heap usage comparison w/ and w/o the change, esp. under stress testing and/or longevity testing. I think these numbers belong to this document to better understand the performance impact.

Reply:
At the design proposal level, all that can be noted is that there will be more Java allocations. Exact numbers will be only available, after the feature is fully implemented and tested.

Reply:
The only number I can share so far is from my local branch, where this feature is implemented (not the final version). In crypto transfer benchmark, total Java allocations are about 40% higher than in `develop`. Note that the benchmark may or may not reflect the real Hedera services node object allocation patterns.

Reply:
Got it. Thanks! The 40% extra Java allocations still looks somewhat concerning. However, I understand that we won't have actual and complete data until we have the implementation ready. In that case, I suggest to evaluate the heap usage data from various performance tests **before** merging the final implementation into `develop`, and only do so if the performance testing results look acceptable.

IMO, some 10%, or maybe up to 20% of extra heap usage may be acceptable. Anything higher than that would need to be thoroughly studied to ensure the new behavior doesn't negatively impact normal node operations (transactions handling and such) in stress and longevity tests. I suggest to add this remark to this document so that we remember to perform this evaluation while we're implementing and testing this feature, and have a rough idea about what we find acceptable.

Reply:
Let's not mix heap allocations with heap usage together. 40% extra allocations may be OK if they are for short lived objects (as seems to be the case here) and the rate is not too high, i.e. if all is within the GC capacity. 40% extra heap usage (objects surviving multiple garbage collections) would be unacceptable.

Reply:
Yes, I understand the difference. Thanks for adding the clarification.

However, my point still stands: the document should mention a need for performing a comprehensive performance analysis using stress and longevity tests, specifically in the area of heap usage, to ensure the normal node operations are not impacted negatively before the code change makes it into production.

Reply:
Yes, this is what `Performance` section is for. It explicitly lists two areas where Java heap usage and/or allocations may be different from the current implementation. Exact numbers can be provided only after implementation is complete, and a number of real world tests / benchmarks are run. If the numbers were available right away, they would be in the proposal